src/kerberos/mod.rs

```diff
@@ -653,40 +654,225 @@ impl<'a> Kerberos {
+            generate_tgs_req(GenerateTgsReqOptions {
+                realm: &as_rep.0.crealm.0.to_string(),
+                service_principal,
+                session_key: &session_key_1,
+                ticket: as_rep.0.ticket.0,
+                authenticator: &mut authenticator,
+                additional_tickets: None.map(|ticket| vec![ticket]),
+                enc_params: &self.encryption_params,
+                context_requirements: builder.context_requirements,
+            })?;
+
+        let response = self.send(yield_point, &serialize_message(&tgs_req)?).await?;
+
+        // first 4 bytes are message len. skipping them
+        let mut d = picky_asn1_der::Deserializer::new_from_bytes(&response[4..]);
+        let tgs_rep: KrbResult<TgsRep> = KrbResult::deserialize(&mut d)?;
+        let tgs_rep = tgs_rep?;
+
+        info!("TGS exchange finished successfully");
+
+        let session_key_2 = extract_session_key_from_tgs_rep(&tgs_rep, &session_key_1, &self.encryption_params)?;
+
+        self.encryption_params.session_key = Some(session_key_2);
+
+        let seq_num = self.next_seq_number();
+
+        let enc_type = self
+            .encryption_params
+            .encryption_type
+            .as_ref()
+            .unwrap_or(&DEFAULT_ENCRYPTION_TYPE);
+        let authenticator_sub_key = generate_random_symmetric_key(enc_type, &mut OsRng);
+
+        let authenticator_options = GenerateAuthenticatorOptions {
+            kdc_rep: &tgs_rep.0,
+            seq_num: Some(seq_num),
+            sub_key: Some(EncKey {
+                key_type: enc_type.clone(),
+                key_value: authenticator_sub_key,
+            }),
+            checksum: Some(ChecksumOptions {
+                checksum_type: AUTHENTICATOR_CHECKSUM_TYPE.to_vec(),
+                checksum_value: AUTHENTICATOR_DEFAULT_CHECKSUM.to_vec(),
+            }),
+            channel_bindings: self.channel_bindings.as_ref(),
+            extensions: Vec::new(),
+        };
+
+        let authenticator = generate_authenticator(authenticator_options)?;
+        let encoded_auth = picky_asn1_der::to_vec(&authenticator)?;
+        info!(encoded_ap_req_authenticator = ?encoded_auth);
+
+        let mech_id = oids::krb5();
+
+        let context_requirements = builder.context_requirements;
+
+        let ap_req = generate_ap_req(
+            tgs_rep.0.ticket.0,
+            self.encryption_params.session_key.as_ref().unwrap(),
+            &authenticator,
+            &self.encryption_params,
+            context_requirements.into(),
+        )?;
+
+        let generated_neg_ap_req = generate_neg_ap_req_plain(ap_req, mech_id)?;
+        let encoded_neg_ap_req = picky_asn1_der::to_vec(&generated_neg_ap_req)?;
+        let output_token = SecurityBuffer::find_buffer_mut(builder.output, SecurityBufferType::Token)?;
+        // the out put token is already a certificate instead of a contentInfo here, need to figure out why
```

Comment:
Maybe add a `TODO` comment, and create a ticket if this is an important detail that must be tracked.
```suggestion
        // TODO: the out put token is already a certificate instead of a contentInfo here, need to figure out why
```